### ISSUE TITLE
Logistic regression probabilities discrepancy between scikit-learn and jpmml-evaluator

### ISSUE THREAD

**Issue by mmakowski** [jpmml/jpmml-evaluator#82]
I observed significant differences between probabilities output by the pipeline in scikit-learn and the same pipeline loaded from PMML and evaluated with jpmml-evaluator. Comparison of scikit-learn and jpmml-evaluator positive probabilities for a few examples:

```
scikit-learn | jpmml-evaluator
-------------+------------------
    0.000514 | 0.001907
    0.001248 | 0.115080
    0.000528 | 0.931263
    0.000111 | 0.200974
    0.003645 | 0.003482
```

Unfortunately I cannot share the dataset or trained model. The scikit-learn pipeline is constructed as follows:

``` python
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.linear_model import LogisticRegression
from sklearn2pmml import PMMLPipeline, sklearn2pmml
from sklearn2pmml.feature_extraction.text import Splitter as PMMLSplitter

training_set = # ...

pipeline = PMMLPipeline([
    ('vectorizer', CountVectorizer(analyzer='word',
                                   tokenizer=PMMLSplitter(),
                                   preprocessor=None,
                                   ngram_range=(1, 3),
                                   min_df=5,
                                   stop_words='english')),
    ('selector',   SelectKBest(score_func=chi2, k=2000)),
    ('classifier', LogisticRegression())
])
pipeline.fit(training_set['test'], training_set['target_label'])
sklearn2pmml(pipeline, 'pipeline.pmml', with_repr=True)
```

And it is evaluated in Scala as follows:

``` scala
import java.io.FileInputStream
import org.jpmml.evaluator.ModelEvaluatorFactory
import scala.collection.JavaConverters._

val text = // ...

val pmml = org.jpmml.model.PMMLUtil.unmarshal(new FileInputStream("pipeline.pmml"))
val pipeline = ModelEvaluatorFactory.newInstance.newModelEvaluator(pmml)
val input = pipeline.getInputFields.get(0).getName
val outputMap = pipeline.evaluate(Map(input -> text).asJava).asScala
```

I have earlier evaluated a similar pipeline but with `SVC` instead of `LogisticRegression` and the results matched between scikit-learn and jpmml-evaluator, so I suspect the problem lies somewhere in the LR classifier.

versions used:

* scikit-learn 0.18.1
* sklearn2pmml 0.26.0
* jpmml-evaluator 1.3.10

**Comment by vruusmann** [comment]
I am testing the conversion of Scikit-Learn's "bag of words" text models here, and my test for the `LogisticRegressionCV` model type passes with absolute/relative error that is lower than 1e-13:
https://github.com/jpmml/jpmml-sklearn/blob/master/src/test/resources/main.py#L362-L386
https://github.com/jpmml/jpmml-sklearn/blob/master/src/test/java/org/jpmml/sklearn/ClassifierTest.java#L235-L243

Also, the `LogisticRegression` model type is the simplest one (ie. basically an array of coefficients), which makes the suggestion that its conversion might be flawed even more implausible.

I suspect that there is something special with your dataset, and the behaviour of the `sklearn2pmml.feature_extraction.text.PMMLSplitter` transformation differs between Scikit-Learn and (J)PMML environments.

Can you help me to triangulate the problem by running the following experiments:
1. How many data records are there in the dataset, and how many of them are failing? Are we talking about 10%, 1%, 0.1% or 0.01% failure rate? You can use my `org.jpmml.evaluator.TestingExample` command-line application to do it: https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-evaluator-example/src/main/java/org/jpmml/evaluator/TestingExample.java
2. If you carefully observe the failing data records, then do they have something in common? Some common word, sequence of words (ie. N-gram), special symbols, whitespace etc?
3. If you execute the exact same dataset first with `LogisticRegression` and then using some other model type (eg. `SVC`), then does the latter give you correct predictions?
4. Is your classification problem a binary or multi-class one? If it's a multi-class one, then can you "fix" it by (temporarily-) rearranging it as a binary one?

**Comment by mmakowski** [comment]
Thank you for prompt response, and for the guidance! With regards to your questions:

1. Less than 1% of the test examples (592 out of 68683) have score differences smaller than 0.0001. The only cases I have seen the scores to match exactly were those that did not contain any features. In those case both scikit-learn and jpmml-evaluator output the probability correspondig to sigmoid(intercept).
2. It appears that the longer the text (and, presumably, more features) of the example, the larger the discrepancy is. The input text is normalised, there are no symbols, only words in lowercase separated by spaces; virtually all of the characters are ASCII. The same text is provided as input to both scikit-learn and jpmml-evaluator.
3. I tried the same pipeline but with `SVC` instead of `LogisticRegression`, and again see completely different results between scikit-learn and jpmml-evaluator, so I believe your suspicion that it is not LR that is at fault here, is correct.
4. It is a binary classification problem.

After further experiments it appears that the problem occurs when the training set is large. In the case of the data I work with, using 40,000 training examples produces a pipeline with scores matching between the two implementations, while 60,000 training examples results in the discrepancies. I am going to investigate further whether it is the volume of data, or a specific example that causes the problem. Since it is not clear at this point whether the issue is with scikit-learn, sklearn2pmml or jpmml-evaluator, I will close this ticket for now, and will reopen if the evidence points to jpmml-evaluator.
